apps/els_lsp/src/els_refactorerl_utils.erl

```diff
@@ -0,0 +1,197 @@
+      try_connect_node({validate, Node});
+
+    disabled ->
+      notification("RefactorErl is not configured!"),
+      {error, disabled};
+
+    _ ->
+      {error, other}
+  end.
+
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+-spec disable_node(atom()) -> atom().
+%%@doc
+%% Disables the node given in paramter, also notifes the user.
+disable_node(Node) ->
+  Msg = "RefactorErl is disconnected!
+  Reload ELS after you fixed theRefactorErl node!",
+  notification(Msg, ?MESSAGE_TYPE_ERROR),
+
+  els_config:set(refactorerl, #{"node" => {Node, disabled}}),
+  {error, disabled}.
+
+
+  %%@doc
+  %% Runs a Query on the RefactorErl node, returns its result if successfull
+  %% If error happens, it just returns an empty list, as most of times the
+  %% result then is directly converted to POIs.
+  %% If the node timeouts or badrpc will come back, it disables the node
+-spec query(string()) -> list() | {error, disabled}.
+query(Query) ->
+  case referl_node() of
+    {error, _} ->
+      [];
+    {ok, Node} ->
+      P = [{positions, linecol}, {output, msg}],
+      Response = rpc:call(Node, refusr_sq, run, [P, [], Query], maxtimeout()),
+      case Response of
+        {badrpc, _} ->
+          disable_node(Node); % Returns disabled
+        error ->
+          [];
+        _ ->
+          Response
+      end
+  end.
+
+-spec convert_to_poi(any()) -> poi().
+%%@doc
+%% Convert a RefactorErl result to a POI, for the LS system
+%% The RefactorErl format is how the refusr_sq:run\3 returns
+convert_to_poi(ReferlResult) ->
+case ReferlResult of
+    [{_, _, _, DataList}] ->
+      convert_to_poi(DataList);
+    [{{_, {FromLine, FromCol}, {ToLine, ToCol}}, Name} | Tail] ->
+      Range = #{ from => {FromLine, FromCol}, to => {ToLine, ToCol} },
+      Id = refactorerl_poi, %"{module(), 'atom()', 'arity()''}",
+      Poi = els_poi:new(Range, application, Id, Name),
+      [ Poi | convert_to_poi(Tail) ];
+    _ ->
+        [] %TODO Robi notify
+end.
+
+
+%%@doc
+%% Calls the RefactorErl node, to check if it's alive using ri:ls()
+-spec check_node(atom()) -> {error, timeout}
+                          | {error, badrpc}
+                          | {error, other}
+                          | atom().
+check_node(Node) ->
+  Response = rpc:call(Node, ri, ls, [], 10000),
+  case Response of
+    {{ok, _}, {error, _}} ->
+      Node;
+    ok ->
+      Node;
+    {badrpc, timeout} ->
+      {error, timeout};
+    {badrpc, _} ->
+      {error, badrpc};
+    _ ->
+      {error, other}
+  end.
+
+%%@doc
+%% Tries to connect to a node
+%% When statues is validate, it reports the success, and failure as well,
+%% when retry, it won't report.
+-spec try_connect_node({validate | retry, atom()}) -> {error, disconnected}
+                                                     | atom().
+try_connect_node({Status, Node}) ->
```

Comment:
A lot of the code here is related to the handling of both transient and permanent failures for the RefactorErl node. As a first iteration, I would suggest we just assume the node is up and running and log eventual RPC failures.
After all, a RPC error is probably due to a misconfiguration and will most likely require human intervention. What do you think? That should significantly simplify things.

Reply:
Yes, it will need human intervention, but if everything is fine only the first two cases will match, but I would like to keep the others there, to make it a bit more error "resistant", also it helps me a lot during development. So I would like to leave it there, but I do want to re-think in the future when new features will be added, etc...